### PR TITLE
Support non-type template parameters in arrays

### PR DESCRIPTION
**Pull request by aneshlya**
This is continuation of non-type template parameters support. This PR adds support of template parameters in arrays.
Related to #2522 and https://github.com/ispc/ispc/issues/2737

### CODE REVIEW

**Review by nurmukhametov — APPROVED** [review]
LGTM!

File: src/type.h [hunk 1]
```diff
@@ -626,6 +626,8 @@ class ArrayType : public SequentialType {
                             example).
      */
     ArrayType(const Type *elementType, int numElements);
+    ArrayType(const Type *elementType, Symbol *num);
```
Comment (dbabokin) [inline src/type.h:635]: The different type of element count argument probably requires some comments about their intended usage scenarios and meaning.

File: src/type.h [hunk 2]
```diff
@@ -127,7 +127,13 @@ class Type : public Traceable {
     bool IsNumericType() const { return IsFloatType() || IsIntType(); }
 
     /** Returns true if the type is any kind of dependent type. */
-    bool IsDependentType() const;
+    bool IsDependent() const;
+
+    /** Returns true if the type is count dependent (based on template non-type parameter). */
+    bool IsCountDependent() const;
```
Comment (dbabokin) [inline src/type.h:133]: The term "count dependent" is not a standard thing. Hence it requires a definition. I would assume that's only a vector and array types with a count parameter dependent on the non-type template parameter (so things like non-type template parameters being used in other contexts are not included). But dichotomy "type dependent" vs "count dependent" suggests that expressions that use non-type template parameters are also "count dependent".

I.e.
```cpp
template <typename T, int N> int foo(T in) {
    int my_var = N - 1; // is "N-1" count dependent?
    T array[N];
    array[my_var] = 123;
    return array[my_var];
}
```

File: src/type.h [hunk 3]
```diff
@@ -672,7 +672,8 @@ class ArrayType : public SequentialType {
 
     int GetElementCount() const;
     const Type *GetElementType() const;
-
+    /* Returns true if array is unsized*/
```
Comment (dbabokin) [inline src/type.h:675]: Again, would be give some definition of "unsized".

File: src/type.h (hunk 2 above)
Comment (aneshlya) [reply src/type.h:138]: Done

File: src/type.h (hunk 1 above)
Comment (aneshlya) [reply src/type.h:653]: Done

File: src/type.h (hunk 3 above)
Comment (aneshlya) [reply src/type.h]: Done

**Review by dbabokin — APPROVED** [review]
Looks good to me

**Pull request merged**